rmb-sdk-go/peer/router.go

```diff
@@ -0,0 +1,133 @@
+package peer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go"
+	"github.com/threefoldtech/tfgrid-sdk-go/rmb-sdk-go/peer/types"
+)
+
+var (
+	// ErrFunctionNotFound is an err returned if the handler function is not found
+	ErrFunctionNotFound = fmt.Errorf("function is not found")
+)
+
+// Handler is a handler function type
+type HandlerFunc func(ctx context.Context, payload []byte) (interface{}, error)
+
+// Middleware is middleware function type
+type Middleware func(ctx context.Context, payload []byte) (context.Context, error)
+
+type Router struct {
+	handlers map[string]HandlerFunc
+	routes   map[string]Router
+	mw       []Middleware
+}
+
+func NewRouter() Router {
+	return Router{
+		handlers: make(map[string]HandlerFunc),
+		routes:   make(map[string]Router),
+	}
+}
+
+// SubRoute add a route prefix to include more sub routes with handler from it
+func (r *Router) SubRoute(prefix string) Router {
+	if strings.Contains(prefix, ".") {
+		panic("invalid subrouter prefix should not have '.'")
+	}
+
+	sub, ok := r.routes[prefix]
+	if ok {
+		return sub
+	}
+
+	router := NewRouter()
+	r.routes[prefix] = router
+	return router
+}
+
+// WithHandler adds a handler function to a router sub command
+func (r *Router) WithHandler(subCommand string, handler HandlerFunc) {
+	if _, ok := r.handlers[subCommand]; ok {
+		panic("handler function is already registered")
+	}
+
+	r.handlers[subCommand] = handler
+}
+
+// Use adds a middleware to the router
+func (r *Router) Use(mw Middleware) {
+	r.mw = append(r.mw, mw)
+}
+
+func (r *Router) Serve(ctx context.Context, peer Peer, env *types.Envelope, err error) {
```

Comment:
There are some issues with implementation:
- The handler has no way to get the "twin" who sent this message. the Router in the other implementation where u copied most of the code also injects extra values for handler (per request) that holds some request information that can be fetched (sender twin id, and even the full envelope) in case it needs information from it.
- The router has NO way to handle requests in parallel. All handling you do is completely sequential (receive a message, handle, it, send response. Means the peer can't call your handler again until you return which make it sequential.

Reply:
Instead when u receive a message you need to first, fill in the ctx with the extra data for the handler. then immediately start a go routine to process the message, and send response back